mlflow/openai/__init__.py

```diff
@@ -0,0 +1,530 @@
+                       path when the model is loaded.
+    :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+
+    :param signature: :py:class:`ModelSignature <mlflow.models.ModelSignature>`
+                      describes model input and output :py:class:`Schema <mlflow.types.Schema>`.
+                      The model signature can be :py:func:`inferred <mlflow.models.infer_signature>`
+                      from datasets with valid model input (e.g. the training dataset with target
+                      column omitted) and valid model output (e.g. model predictions generated on
+                      the training dataset), for example:
+
+                      .. code-block:: python
+
+                        from mlflow.models.signature import infer_signature
+
+                        train = df.drop_column("target_label")
+                        predictions = ...  # compute model predictions
+                        signature = infer_signature(train, predictions)
+    :param input_example: Input example provides one or several instances of valid
+                          model input. The example can be used as a hint of what data to feed the
+                          model. The given example will be converted to a Pandas DataFrame and then
+                          serialized to json using the Pandas split-oriented format. Bytes are
+                          base64-encoded.
+    :param pip_requirements: {{ pip_requirements }}
+    :param extra_pip_requirements: {{ extra_pip_requirements }}
+    :param metadata: Custom metadata dictionary passed to the model and stored in the MLmodel file.
+
+                     .. Note:: Experimental: This parameter may change or be removed in a future
+                                             release without warning.
+    :param  kwargs: Keyword arguments specific to the OpenAI task, such as the ``temperature`` or
+                    or ``top_p`` value to use for chat completion.
+    """
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+    path = os.path.abspath(path)
+    _validate_and_prepare_target_save_path(path)
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+    task = _get_task_name(task)
+
+    if mlflow_model is None:
+        mlflow_model = Model()
+    if signature is not None:
+        mlflow_model.signature = signature
+    elif task == "chat.completions":
+        mlflow_model.signature = ModelSignature(
+            inputs=Schema(
+                [ColSpec(type="string", name="role"), ColSpec(type="string", name="content")],
+            ),
+            outputs=Schema([ColSpec(type="string", name=None)]),
+        )
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+    if metadata is not None:
+        mlflow_model.metadata = metadata
+    model_data_subpath = "model.json"
```

Comment:
```suggestion
    model_data_subpath = "model.yaml"
```

@jinzhang21 maybe we should use yaml here as well?

Reply:
Right, but @sunishsheth2009 mentioned it doesn't work with YAML because Langchain doesn't serialize / format it properly.

Reply:
> Langchain doesn't serialize / format it properly

@sunishsheth2009 Can you elaborate on this?

Reply:
Yes!! So in Langchain, I think its a bug with Langchain, but this is how it stores the `yaml` file
```
_type: !!python/object/apply:langchain.agents.agent_types.AgentType
- zero-shot-react-description
allowed_tools:
- Search
- Calculator
llm_chain:
  _type: llm_chain
  llm:
    _type: openai
    best_of: 1
    frequency_penalty: 0
    logit_bias: {}
    max_tokens: 256
    model_name: text-davinci-003
    n: 1
    presence_penalty: 0
    request_timeout: null
    temperature: 0.0
    top_p: 1
  memory: null
  output_key: text
  prompt:
    _type: prompt
    input_variables:
    - input
    - agent_scratchpad
    output_parser: null
    partial_variables: {}
    template: 'Answer the following questions as best you can. You have access to
      the following tools:
```

but json is stored like this:
```
{
  "llm_chain": {
    "memory": null,
    "verbose": false,
    "prompt": {
      "input_variables": [
        "input",
        "agent_scratchpad"
      ],
      "output_parser": null,
      "partial_variables": {},
      "template": "Answer the following questions as best you can. You have access to the following tools:\n\nSearch: A search engine. Useful for when you need to answer questions about current events. Input should be a search query.\nCalculator: Useful for when you need to answer questions about math.\n\nUse the following format:\n\nQuestion: the input question you must answer\nThought: you should always think about what to do\nAction: the action to take, should be one of [Search, Calculator]\nAction Input: the input to the action\nObservation: the result of the action\n... (this Thought/Action/Action Input/Observation can repeat N times)\nThought: I now know the final answer\nFinal Answer: the final answer to the original input question\n\nBegin!\n\nQuestion: {input}\nThought:{agent_scratchpad}",
      "template_format": "f-string",
      "validate_template": true,
      "_type": "prompt"
    },
    "llm": {
      "model_name": "text-davinci-003",
      "temperature": 0,
      "max_tokens": 256,
      "top_p": 1,
      "frequency_penalty": 0,
      "presence_penalty": 0,
      "n": 1,
      "best_of": 1,
      "request_timeout": null,
      "logit_bias": {},
      "_type": "openai"
    },
    "output_key": "text",
    "_type": "llm_chain"
  },
  "allowed_tools": [
    "Search",
    "Calculator"
  ],
  "_type": "zero-shot-react-description"
}
```


See that the type is incorrect in yaml but it is correct in json. 
Maybe we can create a bug report with Langchain if we decide to go `yaml` with Langchain Agent. 🤔

Reply:
Thanks!

```
_type: !!python/object/apply:langchain.agents.agent_types.AgentType
```

does seem incorrect.

Reply:
Filed https://github.com/hwchase17/langchain/pull/2998